triton/tut/parallel-shared.rst

```diff
@@ -0,0 +1,257 @@
+
+What is shared memory parallelism?
+----------------------------------
+
+In shared memory parallelism a program will launch multiple processes
+or threads so that it can leverage multiple CPUs available in the machine.
+
+**Slurm reservations for both methods behave similarly.** This document
+will talk about processes, but everything mentioned would be applicable to threads
+as well. See section on
+:ref:`difference between processes and threads<threads-vs-processes>`
+for more information on who proceseses and threads differ.
+
+Communication between processes happends via shared memory. This means
+that all processes need to run on the same machine.
+
+.. figure:: /images/parallel-shared.png
+    :width: 80%
+    :align: center
+
+
+Depending on a program, you might have multiple processes (Matlab parallel
+pool, R parallel-library, Python multiprocessing) or have multiple threads
+(OpenMP threads of BLAS libraries that R/numpy use).
+
+Running a typical multiprocess program
+--------------------------------------
+
+Reserving resources for shared memory programs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Reserving resources for shared memory programs is easy: you'll only need to
+specify how many CPUs you want via ``--cpus-per-task=C``-flag.
+
+For most programs using ``--mem=M`` is the correct way to reserve memory,
+but in some cases the amount of memory needed scales with the number of
+processors. This might happen, for example, if each process opens a different
+dataset or runs its own simulation that needs extra memory allocations.
+In these cases you can use ``--mem-per-core=M`` to specify a memory allocation
+that scales with the number of CPUs. We recommend starting with ``--mem=M``
+if you do not know how your program scales.
+
+
+Running an example shared memory parallel program
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For this example, let's consider
+`pi.py <https://github.com/AaltoSciComp/hpc-examples/blob/master/slurm/pi.py>`_-example
+from the
+`hpc-examples <https://github.com/AaltoSciComp/hpc-examples>`_-repository.
+It estimates pi using Monte Carlo methods and can utilize multiple processes for calculating
+the trials.
+
+You can clone the repository with ``git clone https://github.com/AaltoSciComp/hpc-examples.git``.
+
+The script is in the ``slurm``-folder. You can call the script with
+``python pi.py --nprocs=C N``, where ``N`` is the number of iterations to be done by the
+algorithm and ``C`` is the number of processors to be used for the parallel calculation.
+
+Let's run the program with two processes using ``srun``::
+
+  srun --cpus-per-task=2 --time=00:10:00 --mem=1G python pi.py --nprocs=2 1000000
+
+**It is vitally important to notice that the program needs to be told the amount of
+processes it should use.** The program does not obtain this information from the
+queue system automatically. For more information, see the section on
```

Comment:
```suggestion
queue system automatically (a program could, but always assume it doesn't unless specifically
documented and checked). For more information, see the section on
```
Not sure if this is mentioned below or just shouldn't be mentioned.

Reply:
I don't think this is necessary. A program that automatically finds the number of CPUs has been "told" (via code) to look for the environment variables. I would say that adding more text makes the point murkier. The point is that the program needs to understand that not all of the resources available in the machine are necessarily reserved for it .